libraries/standard/mqtt/src/mqtt.c

```diff
@@ -301,6 +301,235 @@ static MQTTStatus_t validatePublishParams( const MQTTContext_t * pContext,
+ */
+static bool matchWildcards( const char * pTopicFilter,
+                            uint16_t topicFilterLength,
+                            uint16_t filterIndex,
+                            const char * pTopicName,
+                            uint16_t topicNameLength,
+                            uint16_t * pNameIndex,
+                            bool * pMatch );
+
+/**
+ * @brief Match a topic name and topic filter allowing the use of wildcards.
+ *
+ * @param[in] pTopicName The topic name to check.
+ * @param[in] topicNameLength Length of the topic name.
+ * @param[in] pTopicFilter The topic filter to check.
+ * @param[in] topicFilterLength Length of topic filter.
+ *
+ * @return `true` if the topic name and topic filter match; `false` otherwise.
+ */
+static bool matchTopicFilter( const char * pTopicName,
+                              uint16_t topicNameLength,
+                              const char * pTopicFilter,
+                              uint16_t topicFilterLength );
+
+/*-----------------------------------------------------------*/
+
+static bool matchWildcardsSpecialCases( const char * pTopicFilter,
+                                        uint16_t topicFilterLength,
+                                        uint16_t filterIndex,
+                                        uint16_t topicNameLength,
+                                        uint16_t nameIndex )
+{
+    bool matchFound = false;
+
+    assert( pTopicFilter != NULL );
+
+    /* Determine if the last character is reached for the topic name, and the
+     * third to last character is reached for the topic filter. */
+    if( ( nameIndex == ( topicNameLength - 1U ) ) &&
+        ( filterIndex == ( topicFilterLength - 3U ) ) )
+    {
+        /* Determine if the topic filter contains "/#" as the last 2 characters.
+         * The '#' wildcard represents the parent and any number of child levels
+         * in the topic name. For example, the filter "sport/#" matches "sport"
+         * as well as "sport/tennis" topics. */
+        matchFound = ( ( pTopicFilter[ filterIndex + 1U ] == '/' ) &&
+                       ( pTopicFilter[ filterIndex + 2U ] == '#' ) ) ? true : false;
+    }
+    else
+    {
+        /* Determine if the last character is reached for the topic name and,
+         * the second to last character for the topic filter. */
+        if( ( nameIndex == ( topicNameLength - 1U ) ) &&
+            ( filterIndex == ( topicFilterLength - 2U ) ) )
+        {
+            /* Determine if the topic filter contains "/+" as the last 2 characters.
+             * This covers the special case of topic matching when the topic name
+             * ends with '/' but the topic filter ends with "/+". Thus, for example,
+             * topic filter "sport/+" matches the "sport/" but not "sport". */
+            matchFound = ( ( pTopicFilter[ filterIndex ] == '/' ) &&
+                           ( pTopicFilter[ filterIndex + 1U ] == '+' ) ) ? true : false;
```

Comment:
Question, what if the filter is "sport/#", and the topic is "sport/"? That should match, right? Is that case handled anywhere? Do we have a unit test for this?

Reply:
Great catch! I verified that this corner case holds true for `#` wildcard within our demo, and have updated the code and added a unit test for the same.